# EPMT Query API

This workbook will illustrate the usage of the EPMT Query API. 

`EPMT` stores job/process/thread data and metadata in a database, be that a local file
or a database server. Access to the database is provided by the EPMT Query API, which
provides functions specifically tailored to the needs of a performance analyst. 

Data from the query functions can be returned in a variety of formats. The most
useful and recommended option is the `pandas` format, which returns a Pandas
dataframe. Other formats include, `terse`, which provides a concise listing, and
`dict`, which returns collections a lists of python dictionaries. We also have a
powerful `orm` return format, that provides an ORM object, which provides for
lazy execution of SQL queries.

The result of this notebook will walk you through the API calls, starting with
the highest-level job queries, followed by more granular operation-level queries,
followed by low-level process and thread queries. As you analyze your data, you will
likely find this hierarchial approach useful in analyzing performance issues.

## Table of Contents

 * [Getting documentation](#getting-docs)
 * [Importing data for the study](#import-data)
 * [Import module](#import-module)
 * [API function categories](#api-categories)
 * [Job Queries](#job-query)
   * [Output format and converting between formats](#output-formats)
   * [Working with ORM objects](#orm-objects)
   * [Job tags](#job-tags)
   * [Job status](#job-status)
   * [Ordering and filtering jobs](#jobs-order-filter)
   * [Failed jobs](#failed-jobs)
   * [Process sums](#proc-sums-field)
   
 * [Operations](#ops)
   * [Selecting processes in an operation](#select-op-procs)
   * [Operation primitive](#operation-primitive)
   * [`get_ops` API call](#get-ops)
   * [Aggregating operation metrics](#op-metrics)
   * [Data-movement v. useful work](#dm-ops)
   * [`get_op_metrics` grouped by tag](#group-by-tag)
   * [CPU-time v. duration](#cpu-time-v-duration)
   * [Operation costs v. total time](#ops-costs)
   * [Operation root(s) (ADVANCED TOPIC)](#op-roots)

 * [Process Queries](#process-query)
   * [Process tags](#process-tags)
     * [Unique process tags in job (ADVANCED TOPIC)](#job-proc-tags)
   * [Filter and ordering](#filter-processes)
   * [Thread metrics aggregation (ADVANCED TOPIC)](#thread-metrics-aggregation)
   * [Process tree and depth (ADVANCED TOPIC)](#proc-tree)
     * [Process tree walk](#process-tree-walk)
   * [Process roots](#proc-roots)
   * [Process histogram](#procs-histogram)
   * [Process Timeline](#timeline)
 * [Thread Query](#thread-query) 
 * [Useful Attributes of Job/Process/Threads](#useful-attributes)
 * [Case Study](#case-study)
 * [Modifying Job Metadata](#modify-job-metadata)
   * [Annotate jobs](#annotate-jobs)
   * [Analyze jobs](#analyze-jobs) 
 * [Deleting Jobs](#delete-jobs)
   * [Delete jobs based on their start/finish times](#delete-jobs-time-based)


## <a name="getting-docs">Getting to the docs</a>

The module functions have embedded documentation in the form of docstrings. You can access it, 
as you would do for any Python module/function:

To get help for all functions in the module, do `help(<module-name)`:
```
help(eq)
```

To get documentation for a specific function, do something like:
```
help(eq.get_jobs)
```

On the command-line you can get a listing of API functions, by doing:
```
$ epmt help api
```

And, details on an individual function, by doing:
```
$ epmt help api <function-name>   # for example "epmt help api get_jobs"
```

## <a name="import-data">Importing the data for this study</a>

This workbook relies on importing the following data. We use an sqlite database 
in this study, but you can use another database such as `postgresql`.
See the `preset_settings` folder to pick up a template of your choice and edit it
if needed. Save the template in the `epmt` folder as `settings.py`

While not required to do so, it's recommended that you start in a fresh database
so as not to affect your existing data. The sqlite database path is controlled
in `settings.py`, and is typically a file in the user `HOME`.

The tar files we need to import are in the `epmt` install area. 
```
# pick the database settings file of your choice
$ cp ../preset_settings/settings_sqlite_localfile.py settings.py

# backup your existing database
# The path might vary depending on your settings.py
# mv ~/EPMT_DB.sqlite ~/EPMT_DB.sqlite.backup

# now import the data
# First we figure out the EPMT install directory, and then import the .tgz from under it
$ EPMT_INSTALL_DIR="`epmt help| grep install_prefix| sed 's/papiex-//'| awk '{print $2}'`epmt"
$ epmt submit $EPMT_INSTALL_DIR/test/data/query_notebook/*.tgz

# check the list of imported jobs
$ epmt list
['625172', '627922', '629337', '633144', '676007', '680181', '685000', '685016', '692544', '693147', '696127', '802954', '804285']
```

<a name="import-module"></a>

In [1]:
# import the query api module
import epmt_query as eq

# pandas and numpy modules will be needed in the examples below
import pandas as pd
import numpy as np

## <a name="api-categories">API Function Categories</a>

The API functions can be broadly put into a few categories based on the hierarchy of data they operate on: 

**Job-level queries**

These functions operate on individual jobs or a collections of jobs. A job can be thought of
as what we submit to a batch system to execute. It's a collection of processes. Job-level
queries include `get_jobs`, `get_job_tags`, `are_jobs_comparable`, and `comparable_job_partitions`.


**Operation-level queries**

These functions operate on the level of operations. An operation can be considered as a collection of processes that share some common tags. A tag is a collection of key/value pairs (a dictionary, if you will), and we associate these tags with processes at the time of execution using the environment variable `PAPIEX_TAGS`.

Consider, a process tagged with `'op:hsmget;op_instance:1;op_sequence:2'`
and another process tagged, with `'op:hsmget;op_instance:2;op_sequence:5'`. Then in terms of the 
most granular operation, the two processes belong to different operations. However, if we consider the
higher-level operation `'op:hsmget'`, then the two processes belong to the same high-level operation.
So an operation is intimately tied to the tag defining it.

Example of operation queries include, `get_ops`, and `get_op_metrics`.


**Process-level queries**

These functions operate on individual processes. A process constitutes of one or more threads. If a process
consists of a single thread, then the thread data is the process data. For multithreaded runs, the process
data is obtained by summation over the inidividual threads. Process-level tend to be more time-consuming, as the number of processes are many orders of magnitude more than the number of jobs. Example of process-level queries include `get_procs`, `job_proc_tags`, `conv_procs`, `rank_proc_tags_keys` and `timeline`.


**Thread-level queries**

At present we have only a single API call for this category: `get_thread_metrics`.


**Reference model calls**

A reference model is a collection of a limited number of jobs or operations. It is used to
serve as reference when determining whether other jobs or operations are inliers or outliers.
Example of such calls include `create_refmodel`, `delete_refmodels` and `get_refmodels`. We
won't be discussing reference model API calls in this notebook. They will be discussed in detail
in the Outliers notebook.


## <a name="job-query">Job Queries</a>

`get_jobs` is the most often used job-level query. It is used to query the database to find jobs
matching some criteria. It usually takes a `tag` and returns a collection of jobs in the format specified by `fmt`. The returned list can be pruned and/or ordered using `fltr`, `limit` and `order`.

You can also pass in one or more jobs as a `jobs` parameter, most often for format conversion.

Let's get started!

In [2]:
# let's get jobs, we use the job tag to select the jobs
jobs = eq.get_jobs(tags='exp_name:ESM4_historical_D151;exp_component:ocean_month_rho2_1x1deg',fmt='terse')
jobs

['625172',
 '627922',
 '629337',
 '633144',
 '676007',
 '680181',
 '685016',
 '692544',
 '693147',
 '696127',
 '802954',
 '804285']

<a name="output-formats"></a>`fmt` can take one of the following values:
 * `terse` -- this returns a list of job ids, no data
 * `pandas` -- this returns a pandas dataframe, with each row representing a job
 * `dict` -- for a list of python dictionaries, with each dict represeting a job
 * `orm` -- EPMT's database layer object for maximum flexibility and speediest queries
 
 `terse` and `orm` queries are the fastest, and should be used if you just want to count
 the number of items, for instance. `pandas` and `dict` take longer, and contain a lot
 more data. If you expect to do computations or vector operatons, then you will prefer `pandas`.
 If you want to iterate over a list, and deal only with ordinary python dicts, then choose
 `dict`.

In [3]:
# above we got a list of job ids. sometimes we want to see more details
# than just the job id. We can use `conv_jobs` to convert between formats
# Or use `get_jobs` to get the specified format directly.
jobs_df = eq.conv_jobs(jobs, fmt='pandas')
display(jobs_df.columns.values)
print(len(jobs_df), 'jobs')
# show the first three rows
jobs_df[:3]

array(['duration', 'updated_at', 'tags', 'info_dict', 'env_dict',
       'cpu_time', 'annotations', 'env_changes_dict', 'analyses',
       'submit', 'start', 'jobid', 'end', 'jobname', 'created_at',
       'exitcode', 'user', 'rchar', 'syscr', 'syscw', 'wchar', 'majflt',
       'minflt', 'rssmax', 'inblock', 'outblock', 'usertime', 'num_procs',
       'processor', 'vol_ctxsw', 'guest_time', 'read_bytes', 'systemtime',
       'time_oncpu', 'timeslices', 'invol_ctxsw', 'num_threads',
       'write_bytes', 'time_waiting', 'all_proc_tags', 'rdtsc_duration',
       'delayacct_blkio_time', 'cancelled_write_bytes',
       'PERF_COUNT_SW_CPU_CLOCK'], dtype=object)

12 jobs


duration                 updated_at  \
0  1.263066e+10 2020-05-01 14:43:03.455826   
1  1.008073e+10 2020-05-01 15:10:24.587673   
2  6.532174e+09 2020-05-01 14:25:41.068333   

                                                tags  \
0  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...   
1  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...   
2  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...   

                                           info_dict  \
0  {'tz': 'US/Eastern', 'status': {'exit_code': 0...   
1  {'tz': 'US/Eastern', 'status': {'exit_code': 0...   
2  {'tz': 'US/Eastern', 'status': {'exit_code': 0...   

                                            env_dict     cpu_time annotations  \
0  {'PWD': '/vftmp/Jeffrey.Durachta/job625172', '...  959968389.0          {}   
1  {'PWD': '/vftmp/Jeffrey.Durachta/job676007', '...  640906121.0          {}   
2  {'PWD': '/vftmp/Jeffrey.Durachta/job627922', '...  679701225.0          {}   

  env_changes_dict analyses                     submit  ... timeslices  \
0               {}       {} 2019-06-09 18:53:22.574059  ...    3175850   
1               {}       {} 2019-06-14 08:30:37.421228  ...     849400   
2               {}       {} 2019-06-10 06:23:14.388744  ...     824685   

  invol_ctxsw num_threads   write_bytes time_waiting  \
0       30049       13471  134643470336  48105200287   
1       12819        3596   73864601600  23712986693   
2       11945        3596   72541306880  13802065001   

                                       all_proc_tags  rdtsc_duration  \
0  [{'op': 'cp', 'op_instance': '1', 'op_sequence...  99908669360503   
1  [{'op': 'cp', 'op_instance': '11', 'op_sequenc...  90008938460454   
2  [{'op': 'cp', 'op_instance': '11', 'op_sequenc...  52576073050078   

   delayacct_blkio_time  cancelled_write_bytes  PERF_COUNT_SW_CPU_CLOCK  
0                     0             2448543744             866184046169  
1                     0             1465577472             605349533801  
2                     0             3044032512             649725614194  

[3 rows x 44 columns]

In [4]:
# if you prefer dealing with python lists and dictionaries,
# you can set fmt='dict'. We recommend, using 'pandas' for scalability
# and the rich API pandas provides.
# Notice we can use `get_jobs` to convert formats
# Below we get a list of dictionaries (showing the first entry only)
eq.get_jobs(jobs = jobs, fmt='dict')[:1]

[{'duration': 12630660818.0,
  'updated_at': datetime.datetime(2020, 5, 1, 14, 43, 3, 455826),
  'tags': {'atm_res': 'c96l49',
   'ocn_res': '0.5l75',
   'exp_name': 'ESM4_historical_D151',
   'exp_time': '18540101',
   'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101',
   'exp_component': 'ocean_month_rho2_1x1deg'},
  'info_dict': {'tz': 'US/Eastern',
   'status': {'exit_code': 0,
    'exit_reason': 'none',
    'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101',
    'script_path': '/home/Jeffrey.Durachta/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/scripts/postProcess/ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101.tags'},
   'process_tree': 1,
   'post_processed': 1},
  'env_dict': {'PWD': '/vftmp/Jeffrey.Durachta/job625172',
   'TMP': '/vftmp/Jeffrey.Durachta/job625172',
   'EPMT': '/home/Jeffrey.Durachta/workflowDB/build//epmt/epmt',
   'HOME': '/home/Jeffrey.Durachta',
   'HOST': 'pp301',
   'LANG': 'en_US',
   'PATH'

### <a name="orm-objects">ORM queries</a>
There is a very useful format called `orm`, which optimizes queries
and it lets you get the underlying Job (or Process) object directly.
ORMs do lazy-evaluation of queries, so many intermediate ORM operations will
return instantenously, and only when certain operations absolutely
need the data, is the SQL actually fired.

We use [SQLAlchemy](https://www.sqlalchemy.org) for the ORM layer.

In [5]:
jobs_orm = eq.get_jobs(jobs, fmt='orm')
jobs_orm.count(), type(jobs_orm)

(12, sqlalchemy.orm.query.Query)

`jobs_orm` above is a `Query` object. The `Query` object can be iterated
over (like a Python list). You can convert it to a list by using the slice
operator -- `[:]`.

In the above example, the query returns immediately, and the actual jobs
are never loaded from the database, since we never access them.

### <a name="job-tags">Job Tags</a>

Each job has a `tags` field that is set by annotating the job with `EPMT_JOB_TAGS`. 
The job tag is a stored as dictionary of key/value pairs. The most common use of the job tag is for selecting
jobs. You can specify the tag either as a dictionary or as a string, with each key/value
pair separated by semicolons. All the key/value pairs must match for a job to be considered
a match.

You might recall that we mentioned tags in the context of operations. Those tags were associated
with individual processes, and are distinct from these job tags. However both are essentially
a dictionary of key/value pairs.

In [6]:
exp_ESM4_historical_D151_jobs = eq.get_jobs(tags='exp_name:ESM4_historical_D151', fmt='orm')

In [7]:
exp_ESM4_historical_D151_jobs.count()

12

In [8]:
for j in exp_ESM4_historical_D151_jobs:
    print(j.jobid, j.tags)

625172 {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time': '18540101', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101', 'exp_component': 'ocean_month_rho2_1x1deg'}
627922 {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time': '18590101', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18590101', 'exp_component': 'ocean_month_rho2_1x1deg'}
629337 {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time': '18640101', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18640101', 'exp_component': 'ocean_month_rho2_1x1deg'}
633144 {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time': '18690101', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18690101', 'exp_component': 'ocean_month_rho2_1x1deg'}
676007 {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time'

As job queries require tags for filtering, you might want to know the *range* of values keys within the job tags take in a collection of jobs. We provide an API call for this purpose:

In [9]:
eq.get_job_tags(exp_ESM4_historical_D151_jobs)

{'atm_res': 'c96l49',
 'ocn_res': '0.5l75',
 'exp_name': 'ESM4_historical_D151',
 'exp_time': {'18540101',
  '18590101',
  '18640101',
  '18690101',
  '18740101',
  '18790101',
  '18840101',
  '18890101',
  '18940101',
  '18990101',
  '19040101',
  '19090101'},
 'script_name': {'ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18590101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18640101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18690101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18740101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18790101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18840101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18890101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18940101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_18990101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_19040101',
  'ESM4_historical_D151_ocean_month_rho2_1x1deg_19090101'},
 'exp_componen

The output says that among all the `exp_ESM4_historical_D151_jobs`, `ocean_res` had a singular value `0.5l75`, the `exp_name` had the same value `ESM4_historical_D151`, but the `exp_time` ranged over a set of values -- `{'18540101','18590101','18640101','18690101','18740101','18790101','18840101','18890101','18940101','18990101','19040101','19090101'}`. The jobs shared the same value for `exp_component` -- `ocean_month_rho2_1x1deg`.

### <a name="job-status">Job Status</a>

The job status contains information about the exit code, signal received by the job.
It also contains some other useful information such as the job script path, etc.

The exit code, below, is taken as the return status of the first process after `epmt start`. 
Usually this would be the parent shell process. The exit code is 
**not the job status from from the batch system**.

In [10]:
eq.get_job_status('625172')

{'exit_code': 0,
 'exit_reason': 'none',
 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101',
 'script_path': '/home/Jeffrey.Durachta/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/scripts/postProcess/ESM4_historical_D151_ocean_month_rho2_1x1deg_18540101.tags'}

### <a name="jobs-order-filter">Ordering and Filtering Jobs</a>

You can use the `order`, `limit`, and `fltr` option with `get_jobs` to sort and filter the job list.
It is advisable to use `limit` when possible, as it sends a `LIMIT` option to the SQL query
and saves database load time. Remember, the database may have millions of jobs, so an unbounded `get_jobs` query with no filtering and no limits will exhaust the physical memory on the system. 

When you need to use unbounded queries, try to use the `orm` format, so you can run a `count` on the
returned ORM object prior to loading the data from the database.

In the example below, we use the `exitcode` field from the `Job` model. 

In [11]:
# some other useful queries might be for instance to order the jobs
# by duration, and getting the top 5
df = eq.get_jobs(jobs, order=eq.desc(eq.Job.duration), fmt="pandas")
df[['jobid', 'tags', 'duration', 'exitcode']]

jobid                                               tags      duration  \
0   625172  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  1.263066e+10   
1   676007  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  1.008073e+10   
2   685016  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  7.005619e+09   
3   629337  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  6.696039e+09   
4   633144  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  6.625638e+09   
5   627922  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  6.532174e+09   
6   680181  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  6.009934e+09   
7   802954  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  3.879024e+09   
8   696127  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  3.676905e+09   
9   693147  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  3.340305e+09   
10  804285  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  3.098385e+09   
11  692544  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'ex...  7.093009e+08   

    exitcode  
0          0  
1          0  
2          0  
3          0  
4          0  
5          0  
6          0  
7          0  
8          0  
9          0  
10         0  
11         0

<a name="failed-jobs"></a>Let's figure out which if any jobs failed.

In [12]:
eq.get_jobs(jobs_orm, fltr=(eq.Job.exitcode != 0), fmt='terse')

[]

So, none of the jobs had a non-zero exit code.

### <a name="proc-sums-field">Aggregation across job processes</a>

Each job object has a `proc_sums` field that aggregates data across the 
processes of the job. The field itself is a dictionary of key/value pairs.
This field is an attribute in the Job object, and when converting from `orm` 
to the other formats, the underlying key/value pairs of the dictionary are made available 
as top-level fields of the `dict` or `pandas` dataframe. `proc_sums` represents aggregates across
the processes of a job:

In [13]:
j = jobs_orm.first()
sorted(j.proc_sums.keys())

['PERF_COUNT_SW_CPU_CLOCK',
 'all_proc_tags',
 'cancelled_write_bytes',
 'delayacct_blkio_time',
 'guest_time',
 'inblock',
 'invol_ctxsw',
 'majflt',
 'minflt',
 'num_procs',
 'num_threads',
 'outblock',
 'processor',
 'rchar',
 'rdtsc_duration',
 'read_bytes',
 'rssmax',
 'syscr',
 'syscw',
 'systemtime',
 'time_oncpu',
 'time_waiting',
 'timeslices',
 'usertime',
 'vol_ctxsw',
 'wchar',
 'write_bytes']

Now, the fields shown above become available in other formats (`dict` and `pandas`) as top-level fields, while the `proc_sums`
field itself is masked.

In [14]:
j_df = eq.get_jobs(j, fmt='pandas')
sorted(j_df.columns.values)

['PERF_COUNT_SW_CPU_CLOCK',
 'all_proc_tags',
 'analyses',
 'annotations',
 'cancelled_write_bytes',
 'cpu_time',
 'created_at',
 'delayacct_blkio_time',
 'duration',
 'end',
 'env_changes_dict',
 'env_dict',
 'exitcode',
 'guest_time',
 'inblock',
 'info_dict',
 'invol_ctxsw',
 'jobid',
 'jobname',
 'majflt',
 'minflt',
 'num_procs',
 'num_threads',
 'outblock',
 'processor',
 'rchar',
 'rdtsc_duration',
 'read_bytes',
 'rssmax',
 'start',
 'submit',
 'syscr',
 'syscw',
 'systemtime',
 'tags',
 'time_oncpu',
 'time_waiting',
 'timeslices',
 'updated_at',
 'user',
 'usertime',
 'vol_ctxsw',
 'wchar',
 'write_bytes']

## <a name="ops">Operations</a>

As a job may span over 50K processes, it's often helpful to select processes belonging to a particular operation. The selection is done using tags, sets of key/value pairs.
In the general case, the collection of processes in an operation forms a **forest**. In the degenerate case of a single root process, the forest will contain a single tree.

### <a name="select-op-procs">Selecting processes in an operation</a>

We can select the processes in an operation by passing a tag to `get_procs`.
You may limit the selection to a single job or multiple jobs using the
`jobs` parameter to `get_procs`.

In [15]:
# below we use the ORM format as we just want a count on the number of processes in the operation
hsmget_op_procs = eq.get_procs(jobs, tags='op:hsmget', fmt='orm')
hsmget_op_procs.count()

27720

### <a name="operation-primitive">The Operation primitive</a>

Using `get_procs` with a tag to select processes in a operation is somewhat
clumsy. The EPMT Query API defines an **Operation** primitive. The `Operation`
API call is passed one or more jobs, and a `tag`. Internally, it calls `get_procs`.
By using the `Operation` primitive, you get aggregated metrics across the
processes constituting the operation in a `proc_sums` attribute. You can specify a granular
operation such as `{'op': 'timavg', 'op_instance': 100, 'op_sequence': 5 }`, or a more
coarse operation, such as `{'op': 'timavg'}`. The important thing to understand is that
all the processes that constitute the operation will share *ALL* the keys/values of the tag.

In [16]:
# you could also select a more granular operation with a more
# specific, tag, such as {'op': 'hsmget', 'op_instance': 10, 'op_sequence': 25}
op = eq.Operation(jobs, {'op': 'hsmget'})
(op.tags, op.processes.count(), op.proc_sums)

({'op': 'hsmget'},
 27720,
 {'vol_ctxsw': 11465035,
  'cancelled_write_bytes': 1290240,
  'read_bytes': 100880384,
  'delayacct_blkio_time': 0,
  'rchar': 6068618970,
  'timeslices': 11577749,
  'outblock': 620041288,
  'wchar': 317449816715,
  'majflt': 245,
  'num_procs': 27720,
  'numtids': 30212,
  'systemtime': 531837341,
  'syscr': 2446648,
  'guest_time': 0,
  'inblock': 197032,
  'syscw': 1674748,
  'processor': 0,
  'time_oncpu': 2556135725081,
  'rssmax': 198999756,
  'minflt': 39641382,
  'PERF_COUNT_SW_CPU_CLOCK': 2242055598562,
  'cpu_time': 2540237921.0,
  'usertime': 2008400580,
  'duration': 87996944703.0,
  'write_bytes': 317461139456,
  'invol_ctxsw': 82453,
  'time_waiting': 172721616270,
  'rdtsc_duration': -28107424259990747})

### <a name='get-ops'>`get_ops` API call</a>
We also have an API call `get_ops` (much like `get_jobs` and `get_procs`) that supports querying for a collection of operations, and multiple output formats, selected using the option -- `fmt`.

In [17]:
dm_ops = eq.get_ops(jobs, tags = ['op:hsmget', 'op:dmput', 'op:cp', 'op:rm', 'op:mv', 'op:untar'], fmt='pandas')
pd.set_option('display.max_colwidth', 150)
dm_ops[['tags','proc_sums', 'start', 'finish', 'duration']]

tags  \
0  {'op': 'hsmget'}   
1   {'op': 'dmput'}   
2      {'op': 'cp'}   
3      {'op': 'rm'}   
4      {'op': 'mv'}   
5   {'op': 'untar'}   

                                                                                                                                               proc_sums  \
0  {'vol_ctxsw': 11465035, 'cancelled_write_bytes': 1290240, 'read_bytes': 100880384, 'delayacct_blkio_time': 0, 'rchar': 6068618970, 'timeslices': 1...   
1  {'vol_ctxsw': 62157, 'cancelled_write_bytes': 3903488, 'read_bytes': 69959680, 'delayacct_blkio_time': 0, 'rchar': 68898988, 'timeslices': 65638, ...   
2  {'vol_ctxsw': 101887, 'cancelled_write_bytes': 52948992, 'read_bytes': 29171712, 'delayacct_blkio_time': 0, 'rchar': 1583622456, 'timeslices': 155...   
3  {'vol_ctxsw': 16963, 'cancelled_write_bytes': 22139564032, 'read_bytes': 57344, 'delayacct_blkio_time': 0, 'rchar': 119740938, 'timeslices': 24518...   
4  {'vol_ctxsw': 141996, 'cancelled_write_bytes': 286720, 'read_bytes': 373248, 'delayacct_blkio_time': 0, 'rchar': 13787477891, 'timeslices': 147118...   
5  {'vol_ctxsw': 26400, 'cancelled_write_bytes': 573440, 'read_bytes': 1825456128, 'delayacct_blkio_time': 0, 'rchar': 14556202373, 'timeslices': 399...   

                       start                     finish      duration  
0 2019-06-09 18:53:24.294728 2019-06-22 17:43:35.235059  8.799694e+10  
1 2019-06-09 18:53:22.610123 2019-06-22 17:50:01.516486  7.034780e+10  
2 2019-06-09 22:17:47.314092 2019-06-22 17:48:32.888042  5.531310e+08  
3 2019-06-09 22:18:09.621773 2019-06-22 17:49:59.287992  4.702161e+07  
4 2019-06-09 22:18:43.539051 2019-06-22 17:49:59.331484  9.928124e+08  
5 2019-06-09 22:15:34.974112 2019-06-22 17:48:29.733792  9.993222e+07

The dataframe contains one row per `tag` (or operation). `proc_sums` contains aggregate metrics across the underlying processes constituting the operation.

If you think of an operation as a collection of processes sharing a tag, you will realize that an operation
need not have a single `start` and `finish`. Instead, an operation may have many `start` and `finish`, each constituting an `interval`. By default, `get_ops`, to reduce computational complexity, does not determine
all the intervals of an operation. However, if you pass an option, `full = True`, `get_ops` will return
additional fields such as `intervals`, that will be a list of tuples, with each tuple representing a single
interval in the form (`start_time`, `finish_time`). The set of intervals of operation determines all the
periods of execution of the operation. In any operation, a boolean field -- `contiguous` -- indicates whether the
job executed over multiple intervals or a single interval.

In [18]:
ops = eq.get_ops(jobs, tags = ['op:hsmget'], full=True, fmt='pandas')
ops[['jobs', 'tags', 'intervals', 'start', 'finish', 'contiguous', 'duration', 'proc_sums', 'num_runs', 'processes']]

jobs  \
0  [625172, 676007, 627922, 629337, 633144, 680181, 685016, 692544, 693147, 696127, 802954, 804285]   

               tags  \
0  {'op': 'hsmget'}   

                                                                                                                                               intervals  \
0  ((2019-06-09 18:53:24.294728, 2019-06-09 19:04:32.832405), (2019-06-09 19:04:32.843397, 2019-06-09 19:04:32.846370), (2019-06-09 19:04:32.859709, ...   

                       start                     finish  contiguous  \
0 2019-06-09 18:53:24.294728 2019-06-22 17:43:35.235059       False   

       duration  \
0  8.799694e+10   

                                                                                                                                               proc_sums  \
0  {'vol_ctxsw': 11465035, 'cancelled_write_bytes': 1290240, 'read_bytes': 100880384, 'delayacct_blkio_time': 0, 'rchar': 6068618970, 'timeslices': 1...   

   num_runs  \
0      1337   

                                                                                                                                               processes  
0  [{'updated_at': 2020-05-01 15:10:24.591081, 'tags': {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '1'}, 'exename': 'uberftp', 'exitcode': 0,...

As you can see the `hsmget` operation contains `1337` distinct intervals of execution, with no overlapping. This means between `hsmget` operation invocations, other processes (with different tags) executed.

### <a name="op-metrics">Aggregating operation metrics</a>

The `Operation` primitive provides an easy way to obtain aggregates on metrics across
processes in an operation. Before `Operation`, the way to obtain metrics was to
use the `get_op_metrics` API call:

In [19]:
# widen width of column display width to show full tag
pd.set_option('display.max_colwidth', 200)

# get the operations with the top cpu_time summed across all processes. 
# Note, cpu_time is better measure of time spent in an operation than 
# 'duration', which might end up double-counting as in a 
# parent-child process scenario, where the parent waits on the time child.
ops_df = eq.get_op_metrics(['629337', '680181'], fmt='pandas').sort_values(by='cpu_time', ascending=False)
ops_df[['jobid', 'tags', 'duration', 'cpu_time']][:10]

jobid                                                        tags  \
24   629337  {'op': 'fregrid', 'op_instance': '7', 'op_sequence': '80'}   
25   680181  {'op': 'fregrid', 'op_instance': '7', 'op_sequence': '80'}   
29   680181    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '3'}   
31   680181    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '5'}   
116  629337  {'op': 'ncrcat', 'op_instance': '13', 'op_sequence': '76'}   
117  680181  {'op': 'ncrcat', 'op_instance': '13', 'op_sequence': '76'}   
26   629337    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '1'}   
34   629337    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '9'}   
28   629337    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '3'}   
32   629337    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '7'}   

         duration    cpu_time  
24   6.937625e+07  68409594.0  
25   5.581548e+07  55867500.0  
29   1.348386e+09  53789735.0  
31   1.039550e+09  49622360.0  
116  4.819452e+07  48163676.0  
117  4.651732e+07  44601217.0  
26   6.956620e+08  39960829.0  
34   3.831944e+08  36986292.0  
28   2.833971e+09  35084576.0  
32   2.803218e+08  34164711.0

#### <a name="dm-ops">Data movement operations</a>
The above call was slow to execute and resulted in a lot of operations. The `get_op_metrics` call can take a 
list of tags if one knows the operations one cares about. The pruning using the `tags` argument speeds up
the operation significantly. Let's figure out the time spent
in data movement operations</a> v. useful work.
In the call to `get_op_metrics` below, we pass in the *list of tags* that
represent the data-movement operations. As it's a list of tags, it's like
an OR-operation with the tags.

In [20]:
dm_tags = ['op:hsmget', 'op:cp', 'op:dmget', 'op:gcp', 'op:mv', 'op:untar', 'op:tar', 'op:rm']
dm_ops_df = eq.get_op_metrics(jobs, tags = dm_tags)
# only showing the first 15 rows
dm_ops_df[['jobid', 'tags', 'cpu_time', 'duration', 'num_procs']][:15]

jobid              tags     cpu_time      duration  num_procs
0   625172  {'op': 'hsmget'}  525588783.0  1.614511e+10       8860
1   627922  {'op': 'hsmget'}  151229296.0  7.160026e+09       1713
2   629337  {'op': 'hsmget'}  221437661.0  7.716459e+09       1713
3   633144  {'op': 'hsmget'}  207422750.0  7.509555e+09       1713
4   676007  {'op': 'hsmget'}  187083822.0  1.450817e+10       1713
5   680181  {'op': 'hsmget'}  230162385.0  7.836627e+09       1713
6   685016  {'op': 'hsmget'}  123305670.0  7.585973e+09       1713
7   692544  {'op': 'hsmget'}  190831238.0  1.526509e+09       1713
8   693147  {'op': 'hsmget'}  199442956.0  5.148361e+09       1730
9   696127  {'op': 'hsmget'}  201617665.0  4.595266e+09       1713
10  802954  {'op': 'hsmget'}  143887409.0  4.502378e+09       1713
11  804285  {'op': 'hsmget'}  158228286.0  3.762514e+09       1713
12  625172      {'op': 'cp'}   10866143.0  5.759710e+07       1343
13  627922      {'op': 'cp'}   13160937.0  4.974643e+07       1044
14  629337      {'op': 'cp'}   12796976.0  4.730122e+07       1044

While the query above helps, we would like it to aggregate across jobs by tag. This
is easily accomplished by passing the <a name="group-by-tag">`group_by_tag`</a> 
argument to `get_op_metrics`:

In [21]:
dm_ops_df_grouped = eq.get_op_metrics(jobs, tags = dm_tags, group_by_tag = True)
dm_ops_df_grouped[['tags', 'cpu_time', 'duration', 'num_procs']]

tags      cpu_time      duration  num_procs
0      {'op': 'cp'}  1.256722e+08  5.531310e+08      12827
1  {'op': 'hsmget'}  2.540238e+09  8.799694e+10      27720
2      {'op': 'mv'}  1.427014e+08  9.928124e+08        900
3      {'op': 'rm'}  2.666295e+07  4.702161e+07       2940
4   {'op': 'untar'}  4.575064e+07  9.993222e+07       2513

So, the total cpu time spent in all data-movement operations can be calculated easily.

In [22]:
dm_ops_df_grouped['cpu_time'].sum()/1e6

2881.025086

In [23]:
# total cpu time spent in the jobs
# here we want to get the total cpu time of the job without any operation filtering
s = 0
for j in jobs_orm: s += j.cpu_time
s/1e6

7351.686315

In [24]:
# data-movement as a percentage of total time
round((100*__/_), 2)

39.19

#### <a name="cpu-time-v-duration">cpu time v. duration</a>
So, the data-movement operations take about `39%` of the total cpu time across our jobs.
There is a reason we did not use `duration` for our calculation, and instead we used
`cpu_time` (a.k.a exclusive cpu time). The reason is that while we have an accurate 
value for `duration` for a job (wall-clock time of the job script), when we attempt to
determine the `duration` of an operation, we run the risk of double-counting `duration`.

How can that happen? 

If a process forks and waits for a child to terminate. The `duration` or `wall-clock` 
time will end up getting calculated both for the parent process and the child process. 
Since the `get_op_metrics` call selects processes using tags, it will select the parent and
child processes for the operation and sum their durations, not realizing that in some
cases the parent may have been background waiting for the child to complete (for example
in case of a shell waiting for a spawned process to complete before resuming execution).
`cpu_time` on the other hand is the actual time spent by a process on the cpu, and 
cannot be counted twice in such a scenario.

As an exercise, let's do the above computation with `duration`.

In [25]:
dm_ops_df_grouped['duration'].sum()/1e6

89689.842002

In [26]:
jobs_duration = 0
for j in jobs_orm: jobs_duration += j.duration
jobs_duration/1e6

70284.716974

As you can see, the cumulative duration of the data-movement operations is showing up to be higher than the job duration. That's ridiculous! It happened because we double-counted the `duration` of background-waiting processes.
Fortunately, we have API support in `get_op_metrics` to prevent double-counting. However, it does make the call somewhat slower, and is disabled by defualt. To prevent double-counting of duration, we pass the `op_duration_method` argument to `get_op_metrics`.

In [27]:
dm_ops_df_grouped = eq.get_op_metrics(jobs, tags = dm_tags, group_by_tag = True, op_duration_method = 'sum-minus-overlap')
dm_ops_df_grouped[['tags', 'cpu_time', 'duration', 'num_procs']]

tags      cpu_time      duration  num_procs
0      {'op': 'cp'}  1.256722e+08  2.084467e+08      12827
1  {'op': 'hsmget'}  2.540238e+09  6.486920e+10      27720
2      {'op': 'mv'}  1.427014e+08  2.744846e+08        900
3      {'op': 'rm'}  2.666295e+07  3.644114e+07       2940
4   {'op': 'untar'}  4.575064e+07  6.663244e+07       2513

In [28]:
dm_ops_df_grouped['duration'].sum()/1e6, jobs_duration/1e6

(65455.201435, 70284.716974)

As you can see now the `duration` is now less than the cumulative job duration.

### <a name="ops-costs">Operation metrics as a fraction of cumulative job metrics</a>

You might be wondering if there's a quicker way to determine the above data movement computation, and there is! 

We have an API call `ops_costs` to determine the value of a metric for operation as fraction of the value of the metric for the entire job. The metric can be `cpu_time` or `duration`. The call automatically ensures that `duration` of overlapping processes is not double-counted.

In [29]:
# features be `duration` and/or `cpu_time`. Here we use cpu_time
(dm_percent, dm_ops_df, jobs_cpu_time, dm_agg_df_by_job) = eq.ops_costs(jobs, tags = dm_tags, features = ['cpu_time'])
dm_percent

39.189

In [30]:
# only showing the first 15 rows
dm_ops_df[['jobid', 'tags', 'num_procs', 'cpu_time']][:15]

jobid              tags  num_procs     cpu_time
0   625172      {'op': 'cp'}       1343   10866143.0
1   625172  {'op': 'hsmget'}       8860  525588783.0
2   625172      {'op': 'mv'}         75    6685913.0
3   625172      {'op': 'rm'}        245    1931483.0
4   625172   {'op': 'untar'}       1523   15601214.0
5   676007      {'op': 'cp'}       1044    9318521.0
6   676007  {'op': 'hsmget'}       1713  187083822.0
7   676007      {'op': 'mv'}         75    8654613.0
8   676007      {'op': 'rm'}        245    2154422.0
9   676007   {'op': 'untar'}         90    2409549.0
10  627922      {'op': 'cp'}       1044   13160937.0
11  627922  {'op': 'hsmget'}       1713  151229296.0
12  627922      {'op': 'mv'}         75   16288451.0
13  627922      {'op': 'rm'}        245    2634351.0
14  627922   {'op': 'untar'}         90    3668353.0

In [31]:
# this shows the costs aggregated by jobs (across tags)
dm_agg_df_by_job

jobid  job_cpu_time  dm_cpu_time  dm_cpu_time%
0   625172   959968389.0  560673536.0          58.0
1   676007   640906121.0  209620927.0          33.0
2   627922   679701225.0  186981388.0          28.0
3   629337   623964730.0  257558676.0          41.0
4   633144   621768978.0  239827284.0          39.0
5   680181   571561896.0  261270304.0          46.0
6   685016   427082965.0  140949699.0          33.0
7   692544   593701277.0  223632776.0          38.0
8   693147   594222175.0  233669236.0          39.0
9   696127   607235263.0  226836335.0          37.0
10  802954   511524729.0  162043117.0          32.0
11  804285   520048567.0  177961808.0          34.0

The table above shows the percentage of `cpu_time` spent in data-movement operations by job. While, most of the jobs are in the `30-40%` range, the first job is an outlier. This might be because, being the first in the series, it might need it's data to be fetched from long-term storage. While, the jobs following it, would find their data in the cache.

### <a name="op-roots">Operation Root(s) (ADVANCED TOPIC)</a>

Sometimes we are interested in knowing the root process(es) that spawned the processes
constituting an operation, or more simply "which processes started an operation". `op_roots` takes
a job (or collection of jobs) and an operation tag is parameters, and returns a collection
of processes that are the root process(es) for the operation.

In [32]:
# we could also give a list of jobs
op_roots = eq.op_roots(['629337'], 'op:hsmget', fmt='pandas')
op_roots.sort_values('duration', ascending=False)[['tags', 'exename', 'args', 'pid', 'ppid', 'duration']]

tags exename  \
9    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '3'}    perl   
58  {'op': 'hsmget', 'op_instance': '4', 'op_sequence': '20'}    perl   
44  {'op': 'hsmget', 'op_instance': '4', 'op_sequence': '14'}    perl   
62  {'op': 'hsmget', 'op_instance': '7', 'op_sequence': '22'}    perl   
51  {'op': 'hsmget', 'op_instance': '4', 'op_sequence': '17'}    perl   
..                                                        ...     ...   
32   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '9'}      ls   
11   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '3'}      ls   
91  {'op': 'hsmget', 'op_instance': '7', 'op_sequence': '25'}      ln   
89  {'op': 'hsmget', 'op_instance': '7', 'op_sequence': '25'}    date   
90  {'op': 'hsmget', 'op_instance': '7', 'op_sequence': '25'}      ls   

                                                                                                                                                                                                       args  \
9   /home/fms/local/opt/hsm/1.2.2/bin/hsmget -v -t -a /archive/oar.gfdl.bgrp-account/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/history -p /ptmp/Jeffrey.Durachta/archive/Jeffrey.Dur...   
58  /home/fms/local/opt/hsm/1.2.2/bin/hsmget -v -t -a /archive/oar.gfdl.bgrp-account/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/history_refineDiag -p /ptmp/Jeffrey.Durachta/archive/...   
44  /home/fms/local/opt/hsm/1.2.2/bin/hsmget -v -t -a /archive/oar.gfdl.bgrp-account/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/history_refineDiag -p /ptmp/Jeffrey.Durachta/archive/...   
62  /home/fms/local/opt/hsm/1.2.2/bin/hsmget -v -t -a /archive/oar.gfdl.bgrp-account/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/history -p /ptmp/Jeffrey.Durachta/archive/Jeffrey.Dur...   
51  /home/fms/local/opt/hsm/1.2.2/bin/hsmget -v -t -a /archive/oar.gfdl.bgrp-account/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/history_refineDiag -p /ptmp/Jeffrey.Durachta/archive/...   
..                                                                                                                                                                                                      ...   
32                                   /vftmp/Jeffrey.Durachta/job629337/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/ESM4_historical_D151_18640101/18640101.nc/18640101.ocean_month_rho2.nc   
11                                   /vftmp/Jeffrey.Durachta/job629337/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/ESM4_historical_D151_18640101/18610101.nc/18610101.ocean_month_rho2.nc   
91                              -s /vftmp/Jeffrey.Durachta/job629337/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/ESM4_historical_D151_18640101/18640101.nc/18640101.ocean_month_rho2.nc .   
89                                                                                                                                                                                                      +%s   
90                                   /vftmp/Jeffrey.Durachta/job629337/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/ESM4_historical_D151_18640101/18640101.nc/18640101.ocean_month_rho2.nc   

      pid   ppid      duration  
9   17018  16269  2.586874e+09  
58  24526  16269  9.873306e+08  
44  22302  16269  4.903930e+08  
62  26759  16269  4.446170e+08  
51  23024  16269  4.148404e+08  
..    ...    ...           ...  
32  20766  16269  1.700000e+02  
11  20009  16269  1.670000e+02  
91  27559  16269  1.600000e+02  
89  27557  16269  1.560000e+02  
90  27558  16269  1.520000e+02  

[92 rows x 6 columns]

As you can see that a number of root processes for the `hsmget` operation were identified. 
Here, a root process of an operation is defined as process, whose parent does not share
the same operation `tag`. So in other words the processes list above had parents which
were not tagged with the `hsmget` tag. 

## <a name="process-query">Process Queries</a>

A process query returns a collection of one or more processes. The query is
passed a `jobs` parameter to restrict the process set to those that belong to a
specified set of `jobs`. 

Like the job query, the process query can take `tag`, `fmt`, 
`fltr`, `order` and `limit` to filter and format the output. `order` and `limit` become
particularly important in process queries as each job can have thousands of processes,
and that takes time to load from the database. In the same vein, using `fmt=orm` is a good
idea, in process queries as that minimizes the database overhead in certain cases.

By default, with `order` unset the ordering of processes in the returned collection
is ORM and database dependent.

In [33]:
# If you want to get the processes belonging to a job
# here each row in the pandas dataframe contains one job process
# again, you can use the 'terse' fmt option to get just the list of database ids of the processes
procs = eq.get_procs(['629337'], fmt='pandas')
display(sorted(procs.columns.values))
# remove columns with only None values and see the first 10 rows
procs.replace(to_replace=[None], value=np.nan).dropna(axis=1,how="all")[:10]

['PERF_COUNT_SW_CPU_CLOCK',
 'args',
 'cancelled_write_bytes',
 'cpu_time',
 'created_at',
 'delayacct_blkio_time',
 'depth',
 'duration',
 'end',
 'exename',
 'exitcode',
 'gen',
 'guest_time',
 'host',
 'id',
 'inblock',
 'inclusive_cpu_time',
 'info_dict',
 'invol_ctxsw',
 'job',
 'jobid',
 'majflt',
 'minflt',
 'numtids',
 'outblock',
 'parent',
 'path',
 'pgid',
 'pid',
 'ppid',
 'processor',
 'rchar',
 'rdtsc_duration',
 'read_bytes',
 'rssmax',
 'sid',
 'start',
 'syscr',
 'syscw',
 'systemtime',
 'tags',
 'time_oncpu',
 'time_waiting',
 'timeslices',
 'updated_at',
 'user',
 'usertime',
 'vol_ctxsw',
 'wchar',
 'write_bytes']

updated_at  \
0 2020-05-01 15:39:45.740128   
1 2020-05-01 15:39:45.740136   
2 2020-05-01 15:39:45.740139   
3 2020-05-01 15:39:45.740142   
4 2020-05-01 15:39:45.740145   
5 2020-05-01 15:39:45.740148   
6 2020-05-01 15:39:45.740151   
7 2020-05-01 15:39:45.740154   
8 2020-05-01 15:39:45.740157   
9 2020-05-01 15:39:45.740160   

                                                        tags  exename  \
0   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '1'}  uberftp   
1   {'op': 'hsmget', 'op_instance': '3', 'op_sequence': '2'}  uberftp   
2   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '3'}  uberftp   
3   {'op': 'hsmget', 'op_instance': '3', 'op_sequence': '4'}  uberftp   
4   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '5'}  uberftp   
5   {'op': 'hsmget', 'op_instance': '3', 'op_sequence': '6'}  uberftp   
6   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '7'}  uberftp   
7   {'op': 'hsmget', 'op_instance': '3', 'op_sequence': '8'}  uberftp   
8   {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '9'}  uberftp   
9  {'op': 'hsmget', 'op_instance': '3', 'op_sequence': '10'}  uberftp   

   exitcode                     path     id                              args  \
0         0  /app/gcp/2.3.19/uberftp  66183  pnfs16.princeton.rdhpcs.noaa.gov   
1         0  /app/gcp/2.3.19/uberftp  66291  pnfs14.princeton.rdhpcs.noaa.gov   
2         0  /app/gcp/2.3.19/uberftp  66381  pnfs15.princeton.rdhpcs.noaa.gov   
3         0  /app/gcp/2.3.19/uberftp  66476  pnfs14.princeton.rdhpcs.noaa.gov   
4         0  /app/gcp/2.3.19/uberftp  66566  pnfs17.princeton.rdhpcs.noaa.gov   
5         0  /app/gcp/2.3.19/uberftp  66661  pnfs17.princeton.rdhpcs.noaa.gov   
6         0  /app/gcp/2.3.19/uberftp  66751  pnfs17.princeton.rdhpcs.noaa.gov   
7         0  /app/gcp/2.3.19/uberftp  66844  pnfs17.princeton.rdhpcs.noaa.gov   
8         0  /app/gcp/2.3.19/uberftp  66936  pnfs16.princeton.rdhpcs.noaa.gov   
9         0  /app/gcp/2.3.19/uberftp  67031  pnfs14.princeton.rdhpcs.noaa.gov   

   depth    pid   jobid  ...  systemtime  time_oncpu timeslices  invol_ctxsw  \
0      0  16665  629337  ...       11998    65471955         38            5   
1      0  17002  629337  ...       10998    65460809         36            4   
2      0  19974  629337  ...       14997    66232435         36            4   
3      0  20104  629337  ...       13997    65162538         36            4   
4      0  20206  629337  ...       11998    65432910         36            4   
5      0  20333  629337  ...       13997    65196486         36            4   
6      0  20435  629337  ...        9998    65116991         36            4   
7      0  20595  629337  ...       14997    65172697         36            4   
8      0  20715  629337  ...       12998    65234466         37            5   
9      0  20859  629337  ...       13997    65493454         35            3   

   write_bytes time_waiting rdtsc_duration  delayacct_blkio_time  \
0            0     12954175   651013062474                     0   
1            0     13000814    43924647916                     0   
2            0     12886679   475186758454                     0   
3            0     12790105    27952157168                     0   
4            0     13059849   434461186042                     0   
5            0     12974073    28135968746                     0   
6            0     12822983   464102930072                     0   
7            0     12918471    28456337258                     0   
8            0     12796432   526619789826                     0   
9            0     12711666    28297078704                     0   

   cancelled_write_bytes  PERF_COUNT_SW_CPU_CLOCK  
0                      0                 24131372  
1                      0                 24106615  
2                      0                 24277355  
3                      0                 24159304  
4                      0                 24221463  
5                      0   

You could also pass in more than one job, in which case the returned processes
would be a superset of those across the jobs list. Here we use the `orm` format
to speed the query since we just want a `count` of processes.

In [34]:
procs = eq.get_procs(['629337', '625172'], fmt='orm')
procs.count()

15943

### <a name="process-tags">Process Tags</a>

Each process saves a dictionary of key/value pairs, such as:
`{'op': "ncatted", 'op_instance': 12, 'op_sequence': 159}`

As mentioned before, the process tag is commonly used to filter processes that constitute an **operation** using the `tag` option.

In [35]:
# below we get the processes in an operation that is identified by 'op_sequence=66'
op = eq.get_procs(jobs, tags='op:cp;op_instance:11;op_sequence:66', fmt='pandas')
len(op)

1914

#### <a name="job-proc-tags">Unique process tags in a job (ADVANCED TOPIC)</a>

We can determine the unique set of process tags across all the processes in a job (or collection of jobs) using the
`job_proc_tags` API call.

In [36]:
# suppose you want to figure out the unique set of operations
# across all the jobs of interest. We would pass in our list of jobs
# showing the first 15 entries
eq.job_proc_tags(jobs_orm)[:15]

[{'op': 'cp', 'op_instance': '1', 'op_sequence': '119'},
 {'op': 'cp', 'op_instance': '1', 'op_sequence': '122'},
 {'op': 'cp', 'op_instance': '1', 'op_sequence': '123'},
 {'op': 'cp', 'op_instance': '11', 'op_sequence': '167'},
 {'op': 'cp', 'op_instance': '15', 'op_sequence': '180'},
 {'op': 'cp', 'op_instance': '3', 'op_sequence': '131'},
 {'op': 'cp', 'op_instance': '5', 'op_sequence': '140'},
 {'op': 'cp', 'op_instance': '7', 'op_sequence': '149'},
 {'op': 'cp', 'op_instance': '9', 'op_sequence': '158'},
 {'op': 'dmput', 'op_instance': '1', 'op_sequence': '126'},
 {'op': 'dmput', 'op_instance': '2', 'op_sequence': '190'},
 {'op': 'fregrid', 'op_instance': '1', 'op_sequence': '117'},
 {'op': 'fregrid', 'op_instance': '1', 'op_sequence': '121'},
 {'op': 'fregrid', 'op_instance': '2', 'op_sequence': '132'},
 {'op': 'fregrid', 'op_instance': '3', 'op_sequence': '141'}]

### <a name="filter-processes">Filtering and Ordering Processes</a>

`order`, `limit` and `fltr` should be used where possible to reduce query time.

In [37]:
# now let's say we care about a particular operation. 
# Let's find the processes in the operation, and
# sort them by the cpu_time, and then see the top offenders
ncatted_procs = eq.get_procs(jobs_orm, \
                             tags = {'op': 'ncatted'}, \
                             order=eq.desc(eq.Process.cpu_time), \
                             limit=10, \
                             fmt='pandas')
ncatted_procs[['jobid', 'tags', 'exename', 'duration', 'cpu_time']]

jobid                                                         tags  \
0  680181  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
1  680181  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
2  629337  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
3  693147   {'op': 'ncatted', 'op_instance': '5', 'op_sequence': '41'}   
4  629337   {'op': 'ncatted', 'op_instance': '3', 'op_sequence': '32'}   
5  627922  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
6  696127  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
7  633144  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
8  692544  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
9  693147  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   

   exename  duration  cpu_time  
0  ncatted    1256.0   58990.0  
1   ncdump    1112.0   53991.0  
2  ncatted    1143.0   48992.0  
3  ncatted    1118.0   48992.0  
4  ncatted    1119.0   48991.0  
5  ncatted    1037.0   47992.0  
6  ncatted    1082.0   47992.0  
7  ncatted    1085.0   47991.0  
8  ncatted    1053.0   47991.0  
9  ncatted    1042.0   46991.0

We could have used a more precise tag, such as `{'op': 'ncatted', 'op_sequence': 85}`,
for more granular operation selection. And, maybe an exename, such as `ncatted`.
Let's sort the returned list by decreasing process duration.

In [38]:
procs = eq.get_procs(jobs_orm, tags='op:ncatted;op_sequence:85', \
                     fltr=(eq.Process.exename == "ncatted"), \
                     order=(eq.desc(eq.Process.duration)), \
                     fmt='pandas')
procs[['jobid', 'tags', 'exename', 'duration', 'cpu_time', 'exitcode']]

jobid                                                         tags  \
0   680181  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
1   629337  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
2   633144  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
3   696127  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
4   692544  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
5   693147  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
6   627922  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
7   804285  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
8   676007  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
9   802954  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   
10  685016  {'op': 'ncatted', 'op_instance': '15', 'op_sequence': '85'}   

    exename  duration  cpu_time  exitcode  
0   ncatted    1256.0   58990.0         0  
1   ncatted    1143.0   48992.0         0  
2   ncatted    1085.0   47991.0         0  
3   ncatted    1082.0   47992.0         0  
4   ncatted    1053.0   47991.0         0  
5   ncatted    1042.0   46991.0         0  
6   ncatted    1037.0   47992.0         0  
7   ncatted     588.0   22995.0         0  
8   ncatted     569.0   23995.0         0  
9   ncatted     536.0   21996.0         0  
10  ncatted     456.0   20996.0         0

### <a name="thread-metrics-aggregation">Process contains aggregated thread metrics (ADVANCED TOPIC)</a>

The `pandas` (and the `dict`) formats, in addition to having process-level data in each row, also have fields that represent metrics aggregated across the underlying threads of the process, such, as
`rssmax`, `cpu_time`, and `rchar`. The ORM `Process` object instead has a `threads_sums` attribute, 
which is a dictionary containing the above fields.

In [39]:
procs.columns.values

array(['updated_at', 'tags', 'exename', 'exitcode', 'info_dict', 'path',
       'id', 'args', 'depth', 'pid', 'jobid', 'numtids', 'ppid', 'start',
       'cpu_time', 'pgid', 'created_at', 'end', 'inclusive_cpu_time',
       'sid', 'duration', 'gen', 'job', 'host', 'parent', 'user', 'rchar',
       'syscr', 'syscw', 'wchar', 'majflt', 'minflt', 'rssmax', 'inblock',
       'outblock', 'usertime', 'processor', 'vol_ctxsw', 'guest_time',
       'read_bytes', 'systemtime', 'time_oncpu', 'timeslices',
       'invol_ctxsw', 'write_bytes', 'time_waiting', 'rdtsc_duration',
       'delayacct_blkio_time', 'cancelled_write_bytes',
       'PERF_COUNT_SW_CPU_CLOCK'], dtype=object)

### <a name="proc-tree">Process tree and depth (ADVANCED TOPIC)</a>

Every process in a job has a `depth` parameter that denotes its depth
in the process tree, with the root process having a `depth` of `0`.

As the process tree construction is an expensive process, we have disabled
automatic creation of the process tree during ingestion by setting `lazy_compute_process_tree`
to `True` in `settings.py`. This does mean that the `depth` parameter is
ordinarily left unset in the process ORM, dataframe or dictionaries. 
We automatically compute the process tree if it's needed for example
to determine the root process(es), or operation roots, etc.

If for any reason you would like to have the `depth` parameter populated,
you can call the `compute_process_trees` API call:

In [40]:
eq.compute_process_trees(['629337', '625172'])
procs = eq.get_procs(['629337', '625172'], fmt='pandas', order=eq.desc(eq.Process.depth))
procs[['exename', 'pid', 'depth']][:5]

exename    pid  depth
0            sleep  20339      7
1  globus-url-copy  77388      7
2  globus-url-copy  20338      7
3            sleep  77330      7
4            sleep  20222      7

Above we compute the process trees for two jobs, and then select their processes ordered by
decreasing depth. As you can see the process trees have a maximum depth of `7`.

Let's <a name="process-tree-walk"></a>navigate the process tree!

In [41]:
# now let's walk through the process tree. To make this easy, we use the 'orm' format
# first we compute the process tree as we intend to walk down the tree
eq.compute_process_trees(['676007'])
# let's sort the processes by exclusive cpu time
# You will get a sorted list of ORM objects, let's see the top 10
procs = eq.get_procs('676007', order = (eq.desc(eq.Process.cpu_time)), fmt='orm')[:10]
[p.pid for p in procs]

[5488, 5218, 3238, 4196, 4560, 4036, 4837, 13027, 29936, 3809]

In [42]:
# lets pick up the first
p = procs[0]

In [43]:
p.exename

'fregrid'

In [44]:
p.exename, p.args, p.duration, len(p.children), p.numtids

('fregrid',
 '--standard_dimension --input_mosaic ocean_mosaic.nc --input_file all --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello,thkcello,vo,vmo,vhGM,vhml --output_file out.nc',
 72611586.0,
 0,
 1)

In [45]:
parent = p.parent
parent

Process[76259]

In [46]:
parent.exename, parent.args, parent.pid, len(parent.children), len(parent.descendants)

('tcsh',
 '-f /home/Jeffrey.Durachta/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/scripts/postProcess/ESM4_historical_D151_ocean_month_rho2_1x1deg_18740101.tags',
 27339,
 735,
 3380)

In [47]:
# let's see the aggregate thread metrics for this process
p.threads_sums

{'rchar': 15140041925,
 'syscr': 1741380,
 'syscw': 33346,
 'wchar': 2185067697,
 'majflt': 3,
 'minflt': 9946,
 'rssmax': 58112,
 'inblock': 12968064,
 'outblock': 2141984,
 'usertime': 62258535,
 'processor': 0,
 'vol_ctxsw': 355,
 'guest_time': 0,
 'read_bytes': 6639648768,
 'systemtime': 5995088,
 'time_oncpu': 68265620909,
 'timeslices': 536,
 'invol_ctxsw': 180,
 'write_bytes': 1096695808,
 'time_waiting': 15819937,
 'rdtsc_duration': 251077780608,
 'delayacct_blkio_time': 0,
 'cancelled_write_bytes': 0,
 'PERF_COUNT_SW_CPU_CLOCK': 68221670024}

### <a name="proc-roots">Process Roots</a>

Often we are interested in finding the root process(es) in a **job**. We have two API calls to help facilitate that: `get_roots` and `root`. 

`get_roots` and `root` are different from `op_roots`, which we discussed earlier as the former find the root process(es) in a **job** or **collection of jobs**, while latter finds the root processes in an **operation**.

In [48]:
root_procs = eq.get_roots(['629337'], fmt='orm')
root_procs.count()

32

As you can see we don't have a process tree rooted in a single process but, rather, a
*forest* with multiple root processes. Let's see some details about the root processes: 

In [49]:
# the processes are ordered (by default) in order of start time
for r in root_procs: print(r.exename, r.args, r.pid, r.ppid, r.start)

tcsh -f /home/Jeffrey.Durachta/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/scripts/postProcess/ESM4_historical_D151_ocean_month_rho2_1x1deg_18640101.tags 16269 16268 2019-06-10 09:59:22.061279
uberftp pnfs16.princeton.rdhpcs.noaa.gov 16665 1 2019-06-10 10:04:35.421661
uberftp pnfs14.princeton.rdhpcs.noaa.gov 17002 1 2019-06-10 10:05:29.490003
uberftp pnfs15.princeton.rdhpcs.noaa.gov 19974 1 2019-06-10 10:48:05.268188
uberftp pnfs14.princeton.rdhpcs.noaa.gov 20104 1 2019-06-10 10:48:45.756830
uberftp pnfs17.princeton.rdhpcs.noaa.gov 20206 1 2019-06-10 10:48:53.923892
uberftp pnfs17.princeton.rdhpcs.noaa.gov 20333 1 2019-06-10 10:49:32.234725
uberftp pnfs17.princeton.rdhpcs.noaa.gov 20435 1 2019-06-10 10:49:40.690278
uberftp pnfs17.princeton.rdhpcs.noaa.gov 20595 1 2019-06-10 10:50:20.804751
uberftp pnfs16.princeton.rdhpcs.noaa.gov 20715 1 2019-06-10 10:51:35.387297
uberftp pnfs14.princeton.rdhpcs.noaa.gov 20859 1 2019-06-10 10:52:19.740397
ncks -H -v ocn_mosaic_file mo

In [50]:
# eq.root picks up the very first process by start time
first_root = eq.root(['629337'], fmt='orm')
first_root.exename, first_root.args, first_root.pid, first_root.ppid, first_root.start

('tcsh',
 '-f /home/Jeffrey.Durachta/ESM4/DECK/ESM4_historical_D151/gfdl.ncrc4-intel16-prod-openmp/scripts/postProcess/ESM4_historical_D151_ocean_month_rho2_1x1deg_18640101.tags',
 16269,
 16268,
 datetime.datetime(2019, 6, 10, 9, 59, 22, 61279))

Looking at the `ppid` field we can see that the root processes returned by `get_roots` are actually siblings
spawned by a single process. The `pid` of the parent is shown as `1`. You can [read here more about why the parent process becomes init for orphan processes](https://unix.stackexchange.com/questions/476191/processes-ppid-changed-to-1-after-closing-parent-shell). `eq.root` on the other just returns the first process from the sibling set. 

We recommend you use `get_roots` for completeness. However, in certain situations `root` suffices.

### <a name="procs-histogram">Process Histogram</a>

Often we want determine a histogram for an attribute in the process model. For instance, in a job or job collection, how often did we use a particular `executable`. Or, how many process were multthreaded?

In [51]:
# we can give a single job or a list of jobs
eq.procs_histogram(['629337'], attr='exename')

{'uberftp': 38,
 'ncks': 14,
 'tcsh': 1072,
 'mkdir': 16,
 'modulecmd': 11,
 'test': 13,
 'getopt': 15,
 'touch': 11,
 'cat': 17,
 'sed': 476,
 'ls': 44,
 'perl': 74,
 'head': 58,
 'sort': 28,
 'find': 6,
 'grep': 77,
 'date': 30,
 'ln': 26,
 'ncra': 7,
 'rm': 197,
 'tar': 6,
 'cut': 213,
 'basename': 6,
 'fregrid': 6,
 'mv': 37,
 'ncatted': 10,
 'wc': 2,
 'python': 6,
 'logger': 6,
 'python2.7': 3,
 'bash': 73,
 'readlink': 25,
 'make': 65,
 'ncdump': 56,
 'tr': 17,
 'TAVG.exe': 6,
 'dmget': 6,
 'uuidgen': 84,
 'which': 175,
 'pwd': 27,
 'uptime': 38,
 'grid-proxy-info': 44,
 'arch': 19,
 'list_ncvars.exe': 7,
 'du': 7,
 'globus-url-copy': 22,
 'dmput': 1,
 'id': 6,
 'sleep': 19,
 'df': 65,
 'tail': 65,
 'host': 40,
 'time': 10,
 'chmod': 10}

In [52]:
eq.procs_histogram(['629337'], attr='numtids')

{4: 59, 1: 3346, 2: 7}

So the job contained `3346` single-thread processes. `7` processes had two threads each, and `59` processes had `4` threads each. 

You may want to refer to the [model attributes](#useful-attributes) for a listing of the attributes in the
`Process` model.

### <a name="timeline">Timeline</a>
Sometimes you want to get a timeline of the processes in the order they were spawned

In [53]:
procs = eq.timeline(jobs, fmt='pandas', limit=25)
procs[['exename', 'tags', 'start', 'duration']]

exename                                                       tags  \
0        tcsh  {'op': 'dmput', 'op_instance': '2', 'op_sequence': '190'}   
1        tcsh                                                         {}   
2       mkdir                                                         {}   
3   modulecmd                                                         {}   
4        test                                                         {}   
5   modulecmd                                                         {}   
6        test                                                         {}   
7   modulecmd                                                         {}   
8        perl                                                         {}   
9        perl                                                         {}   
10        cat                                                         {}   
11     python                                                         {}   
12     logger                                                         {}   
13       perl                                                         {}   
14       perl                                                         {}   
15     python                                                         {}   
16        cat                                                         {}   
17       perl                                                         {}   
18       perl                                                         {}   
19     python                                                         {}   
20        cat                                                         {}   
21     logger                                                         {}   
22       test                                                         {}   
23  modulecmd                                                         {}   
24       perl                                                         {}   

                        start      duration  
0  2019-06-09 18:53:22.610123  1.263059e+10  
1  2019-06-09 18:53:22.614091  1.130000e+02  
2  2019-06-09 18:53:22.623899  1.310000e+02  
3  2019-06-09 18:53:22.664680  3.656000e+03  
4  2019-06-09 18:53:22.678745  5.400000e+01  
5  2019-06-09 18:53:22.689498  1.551000e+03  
6  2019-06-09 18:53:22.701312  4.100000e+01  
7  2019-06-09 18:53:22.711901  3.586940e+05  
8  2019-06-09 18:53:22.745150  1.582100e+04  
9  2019-06-09 18:53:22.770251  4.346000e+03  
10 2019-06-09 18:53:22.786898  1.199900e+04  
11 2019-06-09 18:53:22.798254  3.700000e+01  
12 2019-06-09 18:53:22.834906  2.870000e+02  
13 2019-06-09 18:53:22.881285  1.502000e+04  
14 2019-06-09 18:53:22.906282  5.126000e+03  
15 2019-06-09 18:53:22.920834  3.400000e+01  
16 2019-06-09 18:53:22.929308  6.400000e+01  
17 2019-06-09 18:53:22.954068  1.475100e+04  
18 2019-06-09 18:53:22.978944  4.391000e+03  
19 2019-06-09 18:53:22.992821  3.300000e+01  
20 2019-06-09 18:53:23.001001  6.300000e+01  
21 2019-06-09 18:53:23.018895  2.830000e+02  
22 2019-06-09 18:53:23.135546  4.300000e+01  
23 2019-06-09 18:53:23.146589  5.165800e+04  
24 2019-06-09 18:53:23.161146  3.445500e+04

## <a name="thread-query">Thread Query</a>

The thread query requires passing one or more *process primary keys* or `Process`
objects to `get_thread_metrics`. Let's illustrate this with an example, where
we first obtain the <a name="root-process">root process</a> of a job:

In [54]:
# let's find the root process for a particular job
root = eq.root('629337', fmt='orm')
root.pid

16269

In [55]:
root_threads_df = eq.get_thread_metrics(root)
display(root_threads_df.columns.values)
root_threads_df[['process_pk', 'tid', 'usertime', 'systemtime', 'rssmax']]

array(['end', 'pid', 'sid', 'tid', 'args', 'path', 'pgid', 'ppid', 'tags',
       'rchar', 'start', 'syscr', 'syscw', 'wchar', 'majflt', 'minflt',
       'rssmax', 'exename', 'inblock', 'numtids', 'exitcode', 'hostname',
       'outblock', 'usertime', 'processor', 'starttime', 'vol_ctxsw',
       'generation', 'guest_time', 'read_bytes', 'systemtime',
       'time_oncpu', 'timeslices', 'invol_ctxsw', 'num_threads',
       'write_bytes', 'time_waiting', 'rdtsc_duration',
       'delayacct_blkio_time', 'cancelled_write_bytes',
       'PERF_COUNT_SW_CPU_CLOCK', 'process_pk'], dtype=object)

process_pk    tid  usertime  systemtime  rssmax
0       69435  16269    454930      352946    5516

## <a name="useful-attributes">Useful attributes in Job, Process and Thread objects</a>

The following are some useful attributes of the job, process and thread objects.
They are accessible when using the `orm` format. They are available in the 
`pandas` and `dict` formats. There is one important thing to note:

`proc_sums` field of the Job object is masked for `pandas` and `dict` formats
and the underlying keys of the dictionary are exposed at the top-level.

`threads_sums` field of the Process object is masked for `pandas` and `dict` format
and the underlying keys of the dictionary are exposed at the top-level.

### Attributes of the Job ORM Model
 - `duration`: this is the wallclock time in microseconds
 - `cpu_time`: user+system time aggregated across all processes of the job
 - `start`:    start time in microseconds since epoch
 - `end`:      end time in microseconds since epoch
 - `jobid`:    database id for job (unique)
 - `exitcode`: return code from job
 - `tags`:     dict of key/value pairs
 - `processes`:list of processes belonging to job
 - `proc_sums`: aggregates across processes of a job
 

### Attributes of the Process ORM Model
 - `duration`: this is the wallclock time in microseconds
 - `cpu_time`: exclusive user+system time for process (aggregated across it's threads)
 - `inclusive_cpu_time`: user+system time for the process and *all its descendants*
 - `start`:    start time in microseconds since epoch
 - `end`:      end time in microseconds since epoch
 - `tags`:     dict of key/value pairs
 - `threads_df`: json serialized dataframe of process threads (ADVANCED)
 - `threads_sums`: key/value pairs consisting of sums of thread metrics (ADVANCED)
 - `numtids`:  number of threads
 - `exename`
 - `args`
 - `pid`
 - `ppid`
 - `id`:       database ID for process
 - `exitcode`
 - `parent`
 - `children`
 - `ancestors`
 - `descendants`
 
 
### Attributes of the Thread ORM Model
 - `usertime`
 - `systemtime`
 - `user+system`
 - `rssmax`
 - `majflt`
 - `read_bytes`
 - `write_bytes`

## <a name="case-study">Case Study using the API</a>

Let's walk through a contrived example that shows how you may use the 
API at different levels (job, operation and processes).

In [56]:
# let's pick a job 676007, and see what the top-level job query yields us.
# We will choose the pandas dataframe 
eq.get_jobs('676007', fmt='pandas')

duration                 updated_at  \
0  1.008073e+10 2020-05-01 15:10:24.587673   

                                                                                                                                                                                                      tags  \
0  {'atm_res': 'c96l49', 'ocn_res': '0.5l75', 'exp_name': 'ESM4_historical_D151', 'exp_time': '18740101', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18740101', 'exp_component': 'oce...   

                                                                                                                                                                                                 info_dict  \
0  {'tz': 'US/Eastern', 'status': {'exit_code': 0, 'exit_reason': 'none', 'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18740101', 'script_path': '/home/Jeffrey.Durachta/ESM4/DECK/ESM4...   

                                                                                                                                                                                                  env_dict  \
0  {'PWD': '/vftmp/Jeffrey.Durachta/job676007', 'TMP': '/vftmp/Jeffrey.Durachta/job676007', 'EPMT': '/home/Jeffrey.Durachta/workflowDB/build//epmt/epmt', 'HOME': '/home/Jeffrey.Durachta', 'HOST': 'pp...   

      cpu_time annotations env_changes_dict analyses  \
0  640906121.0          {}               {}       {}   

                      submit  ... timeslices invol_ctxsw num_threads  \
0 2019-06-14 08:30:37.421228  ...     849400       12819        3596   

   write_bytes time_waiting  \
0  73864601600  23712986693   

                                                                                                                                                                                             all_proc_tags  \
0  [{'op': 'cp', 'op_instance': '11', 'op_sequence': '66'}, {'op': 'cp', 'op_instance': '15', 'op_sequence': '79'}, {'op': 'cp', 'op_instance': '3', 'op_sequence': '30'}, {'op': 'cp', 'op_instance': ...   

   rdtsc_duration  delayacct_blkio_time  cancelled_write_bytes  \
0  90008938460454                     0             1465577472   

   PERF_COUNT_SW_CPU_CLOCK  
0             605349533801  

[1 rows x 44 columns]

In [57]:
# now get the processes that are part of this job, let's sort them by the inclusive time
# we need to pass in the job id to restrict the query to a particular job
# the inclusive_cpu_time sums all the cpu times of the process and its dependents
# in this case you can see that after the top-level shell process shows up at the top
procs = eq.get_procs('676007', order = (eq.desc(eq.Process.inclusive_cpu_time)), fmt='pandas', limit=10)
procs[['exename', 'duration', 'inclusive_cpu_time', 'cpu_time']]

exename      duration  inclusive_cpu_time    cpu_time
0      tcsh  1.008058e+10         639744323.0    726888.0
1   fregrid  7.261159e+07          68253623.0  68253623.0
2      ncra  8.840326e+07          55002636.0  54978641.0
3      tcsh  4.076242e+07          38443149.0     15997.0
4  TAVG.exe  4.035406e+07          38386164.0  38386164.0
5      tcsh  3.485533e+07          34631728.0     13997.0
6  TAVG.exe  3.459367e+07          34583741.0  34583741.0
7      perl  1.377508e+09          33066879.0    888864.0
8      perl  3.851296e+07          32827920.0    911860.0
9      perl  3.796058e+07          32017044.0    790879.0

In [58]:
# you may also want to see the processes that were the big CPU hogs in the job
# so here we order by decreasing cpu time, see the top-10 using limit
cpu_hogs = eq.get_procs('676007', order = (eq.desc(eq.Process.cpu_time)), fmt='pandas', limit=10)
cpu_hogs[['exename', 'args', 'tags', 'depth', 'cpu_time', 'duration']]

exename  \
0          fregrid   
1             ncra   
2         TAVG.exe   
3         TAVG.exe   
4         TAVG.exe   
5  globus-url-copy   
6         TAVG.exe   
7  globus-url-copy   
8  globus-url-copy   
9         TAVG.exe   

                                                                                                                                                                                                      args  \
0  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file all --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello,th...   
1                    -h -O -t 2 --header_pad 16384 18700101.ocean_month_rho2.nc 18710101.ocean_month_rho2.nc 18720101.ocean_month_rho2.nc 18730101.ocean_month_rho2.nc 18740101.ocean_month_rho2.nc all.nc   
2                                                                                                                                                                                                            
3                                                                                                                                                                                                            
4                                                                                                                                                                                                            
5                                                                                                                       -rst -bs 6M -tcp-bs 6M -p 1 -fast -pp -g2 -dcsafe -cc 4 -vb -df /tmp/.gcp_GUC_Nn_4   
6                                                                                                                                                                                                            
7                                                                                                                       -rst -bs 6M -tcp-bs 6M -p 1 -fast -pp -g2 -dcsafe -cc 4 -vb -df /tmp/.gcp_GUC_Z3Ir   
8                                                                                                                       -rst -bs 6M -tcp-bs 6M -p 1 -fast -pp -g2 -dcsafe -cc 4 -vb -df /tmp/.gcp_GUC_6Sbm   
9                                                                                                                                                                                                            

                                                         tags  depth  \
0  {'op': 'fregrid', 'op_instance': '7', 'op_sequence': '80'}      1   
1  {'op': 'ncrcat', 'op_instance': '13', 'op_sequence': '76'}      1   
2   {'op': 'timavg', 'op_instance': '1', 'op_sequence': '28'}      2   
3   {'op': 'timavg', 'op_instance': '5', 'op_sequence': '46'}      2   
4   {'op': 'timavg', 'op_instance': '7', 'op_sequence': '55'}      2   
5    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '1'}      7   
6   {'op': 'timavg', 'op_instance': '9', 'op_sequence': '64'}      2   
7    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '9'}      7   
8    {'op': 'hsmget', 'op_instance': '1', 'op_sequence': '7'}      7   
9   {'op': 'timavg', 'op_instance': '3', 'op_sequence': '37'}      2   

     cpu_time    duration  
0  68253623.0  72611586.0  
1  54978641.0  88403258.0  
2  38386164.0  40354062.0  
3  34583741.0  34593673.0  
4  30688334.0  30696276.0  
5  30557354.0  33120282.0  
6  30490363.0  30502852.0  
7  30229403.0  30912962.0  
8  29680487.0  30775660.0  
9  28661642.0  29212899.0

In [59]:
# the fregrid process took pretty long. Let's see the fregrid operation in more detail
# lets see which processes took the top *inclusive* cpu time.
# Let's limit the output to the top 10 results
# and let's get a pandas dataframe
fregrid_op = eq.get_procs(['676007'], tags = 'op:fregrid', order=eq.desc(eq.Process.inclusive_cpu_time), limit=10, fmt='pandas')
fregrid_op

updated_at  \
0 2020-05-01 15:10:24.627817   
1 2020-05-01 15:10:24.626568   
2 2020-05-01 15:10:24.627456   
3 2020-05-01 15:10:24.627163   
4 2020-05-01 15:10:24.626282   
5 2020-05-01 15:10:24.626876   
6 2020-05-01 15:10:24.627820   
7 2020-05-01 15:10:24.627166   
8 2020-05-01 15:10:24.626879   
9 2020-05-01 15:10:24.627459   

                                                         tags  exename  \
0  {'op': 'fregrid', 'op_instance': '7', 'op_sequence': '80'}  fregrid   
1  {'op': 'fregrid', 'op_instance': '3', 'op_sequence': '40'}  fregrid   
2  {'op': 'fregrid', 'op_instance': '6', 'op_sequence': '67'}  fregrid   
3  {'op': 'fregrid', 'op_instance': '5', 'op_sequence': '58'}  fregrid   
4  {'op': 'fregrid', 'op_instance': '2', 'op_sequence': '31'}  fregrid   
5  {'op': 'fregrid', 'op_instance': '4', 'op_sequence': '49'}  fregrid   
6  {'op': 'fregrid', 'op_instance': '7', 'op_sequence': '80'}       mv   
7  {'op': 'fregrid', 'op_instance': '5', 'op_sequence': '58'}       mv   
8  {'op': 'fregrid', 'op_instance': '4', 'op_sequence': '49'}       mv   
9  {'op': 'fregrid', 'op_instance': '6', 'op_sequence': '67'}       mv   

   exitcode info_dict  \
0         0      None   
1         0      None   
2         0      None   
3         0      None   
4         0      None   
5         0      None   
6         0      None   
7         0      None   
8         0      None   
9         0      None   

                                                        path     id  \
0  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  76132   
1  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  75089   
2  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  75836   
3  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  75587   
4  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  74840   
5  /home/fms/local/opt/fre-nctools/bronx-14/gfdl/bin/fregrid  75338   
6                                                    /bin/mv  76133   
7                                                    /bin/mv  75588   
8                                                    /bin/mv  75339   
9                                                    /bin/mv  75837   

                                                                                                                                                                                                      args  \
0  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file all --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello,th...   
1  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file annual --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello...   
2  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file annual --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello...   
3  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file annual --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello...   
4  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file annual --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello...   
5  --standard_dimension --input_mosaic ocean_mosaic.nc --input_file annual --interp_method conserve_order1 --remap_file .fregrid_remap_file_360_by_180.nc --nlon 360 --nlat 180 --scalar_field volcello...   
6                                                                                                                                                                                            out.nc all.nc   
7                                                                                                                                               

So the `fregrid` had multiple invocations of the `fregrid` executable, and that and
`mv` took the longest time to complete.

<a name="failed-procs"></a>Let's see if there are any failed processes in our job selection.

In [60]:
# Let's find the failed processes across our jobs subset
failed_procs = eq.get_procs(['676007'], fltr=(eq.Process.exitcode > 0), fmt='orm')
failed_procs.count()

106

That's a large number of failed processes!

In [61]:
# The orm also gives an easy way to navigate the process hierarchy
# Let's use the ORM directly to walk through the job
j = eq.get_jobs('629337', fmt='orm').first()
j

Job['629337']

In [62]:
# Notice we have a Job object. The processes in the job
# are available as j.processes
j.duration, j.cpu_time, j.exitcode, j.tags

(6696039124.0,
 623964730.0,
 0,
 {'atm_res': 'c96l49',
  'ocn_res': '0.5l75',
  'exp_name': 'ESM4_historical_D151',
  'exp_time': '18640101',
  'script_name': 'ESM4_historical_D151_ocean_month_rho2_1x1deg_18640101',
  'exp_component': 'ocean_month_rho2_1x1deg'})

In [63]:
# first we ask for the aggregate metrics for single job
# Here, we don't specify any tags. For single jobs, when
# we don't specify the operation/tags, they are queried from the job
op_sums = eq.get_op_metrics(jobs='629337', fmt='pandas')
display(op_sums.columns.values)
# let's see the first 10 rows
op_sums[['jobid', 'tags', 'duration', 'cpu_time']][:10]

array(['read_bytes', 'delayacct_blkio_time', 'timeslices', 'outblock',
       'majflt', 'systemtime', 'syscr', 'guest_time', 'inblock', 'syscw',
       'time_oncpu', 'minflt', 'PERF_COUNT_SW_CPU_CLOCK', 'invol_ctxsw',
       'vol_ctxsw', 'cancelled_write_bytes', 'rchar', 'wchar',
       'processor', 'rssmax', 'usertime', 'write_bytes', 'time_waiting',
       'rdtsc_duration', 'job', 'jobid', 'tags', 'num_procs', 'numtids',
       'cpu_time', 'duration'], dtype=object)

jobid                                                        tags  \
0  629337      {'op': 'cp', 'op_instance': '11', 'op_sequence': '66'}   
1  629337      {'op': 'cp', 'op_instance': '15', 'op_sequence': '79'}   
2  629337       {'op': 'cp', 'op_instance': '3', 'op_sequence': '30'}   
3  629337       {'op': 'cp', 'op_instance': '5', 'op_sequence': '39'}   
4  629337       {'op': 'cp', 'op_instance': '7', 'op_sequence': '48'}   
5  629337       {'op': 'cp', 'op_instance': '9', 'op_sequence': '57'}   
6  629337    {'op': 'dmput', 'op_instance': '2', 'op_sequence': '89'}   
7  629337  {'op': 'fregrid', 'op_instance': '2', 'op_sequence': '31'}   
8  629337  {'op': 'fregrid', 'op_instance': '3', 'op_sequence': '40'}   
9  629337  {'op': 'fregrid', 'op_instance': '4', 'op_sequence': '49'}   

       duration    cpu_time  
0  8.055334e+06   2078506.0  
1  7.426402e+06   1699557.0  
2  8.307672e+06   2191485.0  
3  7.696808e+06   2229484.0  
4  7.868858e+06   2439456.0  
5  7.946142e+06   2158488.0  
6  6.699475e+09   2943528.0  
7  1.378604e+07  13814897.0  
8  1.386925e+07  13894885.0  
9  1.396966e+07  14014866.0

## <a name="modify-job-metadata">Modifying Job Metadata</a>

Every job stores metadata such as job name, `jobid`, `tags`. Most metadata
fields are not designed to be mutable. However, there are fields such
as `annotations` that are permitted to be modfied using the API.

### <a name="annotate-jobs">Annotate Jobs</a>

Job annotations allow you to store arbitrary key/value pairs
in a persistent manner. They may be retrieved later, modified, and
saved again. There is no semantic meaning associated with annotations
other than what the user ascribes to them. 

In [64]:
eq.get_job_annotations('629337')

{}

In [65]:
eq.annotate_job('629337', {'abc': 100})

{'abc': 100}

In [66]:
eq.get_job_annotations('629337')

{'abc': 100}

In [67]:
eq.annotate_job('629337', {'def': 200})

{'abc': 100, 'def': 200}

In [68]:
eq.get_job_annotations('629337')

{'abc': 100, 'def': 200}

You will notice that `annotate_job` *merges* (or *updates*) the dictionary. 
It doesn't remove existing keys unless you overwrite them with a new key.

In [69]:
eq.annotate_job('629337', {'abc': 500})

{'abc': 500, 'def': 200}

If you wish to replace the existing annotations completely, you should
set the `replace` flag to `annotate_job`.

In [70]:
eq.annotate_job('629337', {'my_key': "something new"}, replace = True)

{'my_key': 'something new'}

In [71]:
eq.get_job_annotations('629337')

{'my_key': 'something new'}

To remove all annotations:

In [72]:
eq.remove_job_annotations('629337')

{}

In [73]:
eq.get_job_annotations('629337')

{}

### <a name="analyze-jobs">Analyze Jobs</a>

We have developed API calls to run simple analyses pipelines, such
as outlier detection on jobs. The output from such analyses is 
persisted a dictionary along with the job in the database, and
can be retrieved later. 

Please note that the current analyses
pipeline is simplistic and will be improved. It is meant solely
for illustrative purposes and to facilitate feedback to spur
subsequent development.

In [74]:
# request list of *all* unanalyzed jobs
eq.get_unanalyzed_jobs()

['625172',
 '627922',
 '629337',
 '633144',
 '676007',
 '680181',
 '685016',
 '692544',
 '693147',
 '696127',
 '802954',
 '804285']

In [75]:
# usually we care about a subset of recent jobs rather than
# querying the whole database
eq.get_unanalyzed_jobs(['625172', '627922','629337','633144'])

['625172', '627922', '629337', '633144']

You may also care about a specific analysis, such as outlier detection
in which case you can pass an `analyses_filter` to define what an 
"analyzed" job means. We will cover this later in an example once we
have some jobs that we have analyzed.

First let's run an analysis pipeline on some jobs:

In [76]:
# here we pass a list of jobs we want to analyze. If don't pass
# an argument (or an empty list) all pending jobs will be analyzed.
# That could take very long. So here we retrict the active set..
num_filters_executed = eq.analyze_pending_jobs(['625172', '627922', '629337', '633144'])

`analyze_pending_jobs` returns the number of analysis methods executed on the
jobs. At present we only have `outlier_detection` in our pipeline. So the
`num_filters_executed` will be `1`. If you run the the same call again, the 
outlier detection algorithm will not be run again. Now, let's see what 
results came from our analysis.

In [77]:
eq.get_job_analyses('625172')

{'outlier_detection': [{'results': {'cpu_time': [['627922',
      '629337',
      '633144'],
     ['625172']],
    'duration': [['627922', '629337', '633144'], ['625172']],
    'num_procs': [['627922', '629337', '633144'], ['625172']]},
   'model_id': None}]}

Since we ran outlier detection on 4 jobs, each of the jobs has the same
analyses saved. We queried one of the jobs. We could have queried one
of the other jobs and got the same output.

In [78]:
eq.get_job_analyses('627922')

{'outlier_detection': [{'results': {'cpu_time': [['627922',
      '629337',
      '633144'],
     ['625172']],
    'duration': [['627922', '629337', '633144'], ['625172']],
    'num_procs': [['627922', '629337', '633144'], ['625172']]},
   'model_id': None}]}

The results suggest that based on each of the features -- `num_procs`,
`duration`, and `cpu_time` -- job `625172` is an outlier, while
the other three jobs `['633144', '629337', '627922']` form an 
equivalent set.

Now if we query the list of unanalyzed jobs, these 4 jobs will be
absent.

In [79]:
eq.get_unanalyzed_jobs()

['676007',
 '680181',
 '685016',
 '692544',
 '693147',
 '696127',
 '802954',
 '804285']

## <a name="delete-jobs">Deleting Jobs</a>

In [80]:
len(eq.get_jobs(fmt='terse'))

12

In [81]:
# to delete a single job
# the function returns the number of jobs deleted
eq.delete_jobs('804285')

1

In [82]:
# to delete multiple jobs, you need to set a second `force` argument
eq.delete_jobs(['693147','696127'], force=True)

2

### <a name="delete-jobs-time-based">Time-based Job Deletion</a>

In [83]:
# to delete all jobs older than 30 days
# Very useful in a cron job!
eq.delete_jobs([], force=True, before=-30)

9

In [84]:
# to delete jobs that were executed in the last week
eq.delete_jobs([], force=True, after=-7)

0

In [85]:
# to delete jobs older than a specific date
# delete jobs before Jan 21, 2018 09:55
eq.delete_jobs([], force=True, before='01/21/2018 09:55')

0